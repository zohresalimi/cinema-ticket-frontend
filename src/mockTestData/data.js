import React from "react";
import { Router } from "@reach/router";
import AppContext from "../store/context";
import Theme from "../Styles/Theme";

export const getTestStore = () => {
  return {
    dispatch: () => {},
    state: {
      user: {
        name: "",
        email: "",
      },
      movies: {
        children: [
          {
            _id: "5fa71823adc463397a50c501",
            actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
            age: "7",
            coverImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/68bfeef55ada4e7ab0606db2af0f40ad.jpg?width=240&version=EA2E80EE161E5DA161C6E26487054C9F&format=webp",
            description:
              "Poppy and Kvist embark on an adventure that will take them far beyond",
            director: "aaron woodley",
            duration: "1:31",
            genre: ["family", "children"],
            images: [],
            largeImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/79276cead82341d2a004f8e8f99a88e9.jpg?width=1920&version=08F0C83177D59FB94C4C809F63F67E2A&format=webp",
            name: "trolls 2: world tour",
            originalTitle: "swedish",
            premiere: "2020-10-03T15:18:52.722Z",
            rooms: ["5fa653d0a7f1201999dd1ac0"],
            trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
          },
        ],
        premiered: [
          {
            _id: "5fa71823adc463397a50c502",
            actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
            age: "7",
            coverImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/68bfeef55ada4e7ab0606db2af0f40ad.jpg?width=240&version=EA2E80EE161E5DA161C6E26487054C9F&format=webp",
            description:
              "Poppy and Kvist embark on an adventure that will take them far beyond",
            director: "aaron woodley",
            duration: "1:31",
            genre: ["family", "children"],
            images: [],
            largeImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/79276cead82341d2a004f8e8f99a88e9.jpg?width=1920&version=08F0C83177D59FB94C4C809F63F67E2A&format=webp",
            name: "follow me",
            originalTitle: "swedish",
            premiere: "2020-10-05T15:18:52.722Z",
            rooms: ["5fa653c0643c06586bba7d78"],
            trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
          },
        ],
        upcoming: [
          {
            _id: "5fa71823adc463397a50c503",
            actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
            age: "7",
            coverImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/68bfeef55ada4e7ab0606db2af0f40ad.jpg?width=240&version=EA2E80EE161E5DA161C6E26487054C9F&format=webp",
            description:
              "Poppy and Kvist embark on an adventure that will take them far beyond",
            director: "aaron woodley",
            duration: "1:31",
            genre: ["family", "children"],
            images: [],
            largeImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/79276cead82341d2a004f8e8f99a88e9.jpg?width=1920&version=08F0C83177D59FB94C4C809F63F67E2A&format=webp",
            name: "honest thief",
            originalTitle: "swedish",
            premiere: "2021-01-03T15:18:52.722Z",
            rooms: ["5fa653c0643c06586bba7d78"],
            trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
          },
        ],
        allMovies: {
          "5fa71823adc463397a50c501": "children",
          "5fa71823adc463397a50c502": "premiered",
          "5fa71823adc463397a50c503": "upcoming",
        },
      },
      cinemas: [
        {
          _id: "5fa64fe234fbbf9945f3a3ce",
          name: "filmstaden kista",
          purchaseEndTime: "21:30",
          purchaseStartTime: "8:00",
          rooms: ["5fa653c0643c06586bba7d78", "5fa653d0a7f1201999dd1ac0"],
        },
      ],
      ticket: {
        movie: {},
        status: false,
        quantity: 0,
        price: 0,
        seatNumbers: new Map(),
        user: {},
        showing: {},
      },
      showings: [
        {
          _id: "5fbe9bd31f58462b5c759b5c",
          capacity: 4,
          cinema: "5fa64fe234fbbf9945f3a3ce",
          endTime: null,
          movie: {
            _id: "5fa71823adc463397a50c502",
            actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
            age: "7",
            description:
              "Poppy and Kvist embark on an adventure that will take them far beyond",
            director: "aaron woodley",
            duration: "1:31",
            genre: ["family", "children"],
            images: [],
            name: "follow me",
            originalTitle: "swedish",
            premiere: "2020-10-05T15:18:52.722Z",
            rooms: ["5fa653c0643c06586bba7d78"],
            trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
          },
          price: 100,
          room: { _id: "5fa653c0643c06586bba7d78" },

          seats: [
            [
              { seatNumber: 1, taken: "available" },
              { seatNumber: 1, taken: "available" },
            ],
            [
              { seatNumber: 1, taken: "available" },
              { seatNumber: 1, taken: "available" },
            ],
          ],
          startTime: "2020-11-13T16:00:00.000Z",
        },
      ],
    },
  };
};

export const WithProvider = (props) => {
  const { state, dispatch } = getTestStore();
  return (
    <Theme>
      <AppContext.Provider
        value={{ state: props.defaultValue || state, dispatch }}
      >
        <Router>{props.children}</Router>
      </AppContext.Provider>
    </Theme>
  );
};
