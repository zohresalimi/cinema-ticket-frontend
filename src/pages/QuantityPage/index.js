import React, { useContext, useEffect } from "react";
import { Link, useLocation } from "@reach/router";
import { useTranslation } from "react-i18next";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faAngleLeft } from "@fortawesome/free-solid-svg-icons";

import AppContext from "../../store/context";
import {
  SET_TICKET_QUANTITY_REDUCER,
  SET_PRICE_REDUCER,
} from "../../constants";
import {
  Container,
  Row,
  Button,
  Breadcrumb,
  Alert,
} from "../../Styles/StyleComponents";
import Quantity from "./style";

function QuantityPage() {
  const location = useLocation();
  const { state, dispatch } = useContext(AppContext);
  const { movie, showing, quantity } = state.ticket;
  const { category } = location.state;
  const { t } = useTranslation();

  const handleIncrement = () => {
    dispatch({ type: SET_TICKET_QUANTITY_REDUCER, data: quantity + 1 });
  };

  const handleDecrement = () => {
    if (quantity > 0) {
      dispatch({ type: SET_TICKET_QUANTITY_REDUCER, data: quantity - 1 });
    }
  };

  useEffect(() => {
    if (showing) {
      dispatch({
        type: SET_PRICE_REDUCER,
        data: showing.price * quantity,
      });
    }
  }, [dispatch, quantity, showing]);

  return (
    <Container>
      <Row>
        <div className="full-width">
          <Breadcrumb>
            <Link to={`/${category}/movie-detail/${movie._id}`}>
              <FontAwesomeIcon icon={faAngleLeft} className="mr-10" />
              {t(movie.name)}
            </Link>
          </Breadcrumb>
          <Quantity>
            <p className="breadcrumb">{t("number of tickets")}</p>
            <div>
              <button data-testid="decrement" onClick={handleDecrement}>
                -
              </button>
              <span data-testid="ticket-quantity">{quantity}</span>
              <button data-testid="increment" onClick={handleIncrement}>
                +
              </button>
            </div>
          </Quantity>
          <Button disabled={quantity <= 0}>
            <Link
              to={`/booking/${showing._id}/seat-selection`}
              state={{ category }}
            >
              {t("select location")}
            </Link>
          </Button>
          {quantity === 0 && (
            <Alert>{t("you must select at least one ticket")}</Alert>
          )}
        </div>
      </Row>
    </Container>
  );
}

export default QuantityPage;
