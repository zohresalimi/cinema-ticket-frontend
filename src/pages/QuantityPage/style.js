import styled from "styled-components";

const Quantity = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  p {
    font-size: ${(props) => props.theme.fontSizes.medium};
  }
  div {
    button {
      width: 30px;
      height: 30px;
      cursor: pointer;
      font-size: ${(props) => props.theme.fontSizes.small};
    }
    span {
      font-family: ${(props) => props.theme.fontFamily.sf};
      font-size: ${(props) => props.theme.fontSizes.medium};
      width: 60px;
      display: inline-block;
      text-align: center;
    }
  }
`;

export { Quantity as default };
