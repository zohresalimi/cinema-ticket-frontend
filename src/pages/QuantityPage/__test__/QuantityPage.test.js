import * as React from "react";
import * as ReachRouter from "@reach/router";

import { act, render, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import { getTestStore, WithProvider } from "../../../mockTestData/data";
import QuantityPage from "../";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({
    t: (k) => k,
    i18n: {
      language: "sv",
      changeLanguage: () => new Promise(() => {}),
    },
  }),
}));

async function renderWrapper(defaultValue, props) {
  let component;

  act(() => {
    component = render(
      <WithProvider defaultValue={defaultValue}>
        <QuantityPage {...props} path="/" />
      </WithProvider>
    );
  });

  return component;
}

describe("Quantity Page Component Testing", () => {
  afterAll(() => {
    jest.clearAllMocks();
  });

  test("take snapshot", async () => {
    const mockUseLocation = jest.fn(() => ({
      state: {
        category: "premiered",
      },
    }));

    const mockUseNavigate = jest.fn();

    jest.spyOn(ReachRouter, "useLocation").mockImplementation(mockUseLocation);
    jest.spyOn(ReachRouter, "useNavigate").mockImplementation(mockUseNavigate);

    const { container } = await renderWrapper();
    expect(container.firstChild).toMatchSnapshot();
  });

  test("should increase quantity", async () => {
    const testState = getTestStore();
    testState.ticket.showing = testState.showings[0];
    const mockUseLocation = jest.fn(() => ({
      state: {
        category: "premiered",
      },
    }));

    const mockUseNavigate = jest.fn();

    jest.spyOn(ReachRouter, "useLocation").mockImplementation(mockUseLocation);
    jest.spyOn(ReachRouter, "useNavigate").mockImplementation(mockUseNavigate);

    const { container, getByTestId } = await renderWrapper(testState);
    const increment = await getByTestId("increment");
    fireEvent.click(increment);
    expect(getByTestId("ticket-quantity")).toHaveTextContent("1");
    expect(container.firstChild).toMatchSnapshot();
  });

  test("should decrease quantity", async () => {
    const mockUseLocation = jest.fn(() => ({
      state: {
        category: "premiered",
      },
    }));

    const mockUseNavigate = jest.fn();

    jest.spyOn(ReachRouter, "useLocation").mockImplementation(mockUseLocation);
    jest.spyOn(ReachRouter, "useNavigate").mockImplementation(mockUseNavigate);

    const testState = getTestStore();
    testState.ticket.showing = testState.showings[0];

    const { container, getByTestId } = await renderWrapper(testState);
    const decrement = await getByTestId("decrement");
    const increment = await getByTestId("increment");
    fireEvent.click(increment);
    fireEvent.click(increment);
    fireEvent.click(increment);
    fireEvent.click(decrement);
    expect(getByTestId("ticket-quantity")).toHaveTextContent("2");
    expect(container.firstChild).toMatchSnapshot();
  });
});
