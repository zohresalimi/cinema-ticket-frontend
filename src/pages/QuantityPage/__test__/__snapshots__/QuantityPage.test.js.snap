// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quantity Page Component Testing should decrease quantity 1`] = `
<div
  style="outline: none;"
  tabindex="-1"
>
  <div
    class="StyleComponents__Container-sc-11erdxt-0 fFgwWO"
  >
    <div
      class="StyleComponents__Row-sc-11erdxt-1 jhoSVF"
    >
      <div
        class="full-width"
      >
        <h2
          class="StyleComponents__Breadcrumb-sc-11erdxt-4 cNjPuO"
        >
          <a
            href="/premiered/movie-detail/undefined"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-angle-left fa-w-8 mr-10"
              data-icon="angle-left"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 256 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
                fill="currentColor"
              />
            </svg>
          </a>
        </h2>
        <div
          class="style__Quantity-wqle2v-0 eTLLsO"
        >
          <p
            class="breadcrumb"
          >
            number of tickets
          </p>
          <div>
            <button
              data-testid="decrement"
            >
              -
            </button>
            <span
              data-testid="ticket-quantity"
            >
              2
            </span>
            <button
              data-testid="increment"
            >
              +
            </button>
          </div>
        </div>
        <button
          class="StyleComponents__Button-sc-11erdxt-3 fWbzoT"
        >
          <a
            href="/booking/undefined/seat-selection"
          >
            select location
          </a>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Quantity Page Component Testing should increase quantity 1`] = `
<div
  style="outline: none;"
  tabindex="-1"
>
  <div
    class="StyleComponents__Container-sc-11erdxt-0 fFgwWO"
  >
    <div
      class="StyleComponents__Row-sc-11erdxt-1 jhoSVF"
    >
      <div
        class="full-width"
      >
        <h2
          class="StyleComponents__Breadcrumb-sc-11erdxt-4 cNjPuO"
        >
          <a
            href="/premiered/movie-detail/undefined"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-angle-left fa-w-8 mr-10"
              data-icon="angle-left"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 256 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
                fill="currentColor"
              />
            </svg>
          </a>
        </h2>
        <div
          class="style__Quantity-wqle2v-0 eTLLsO"
        >
          <p
            class="breadcrumb"
          >
            number of tickets
          </p>
          <div>
            <button
              data-testid="decrement"
            >
              -
            </button>
            <span
              data-testid="ticket-quantity"
            >
              1
            </span>
            <button
              data-testid="increment"
            >
              +
            </button>
          </div>
        </div>
        <button
          class="StyleComponents__Button-sc-11erdxt-3 fWbzoT"
        >
          <a
            href="/booking/undefined/seat-selection"
          >
            select location
          </a>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Quantity Page Component Testing take snapshot 1`] = `
<div
  style="outline: none;"
  tabindex="-1"
>
  <div
    class="StyleComponents__Container-sc-11erdxt-0 fFgwWO"
  >
    <div
      class="StyleComponents__Row-sc-11erdxt-1 jhoSVF"
    >
      <div
        class="full-width"
      >
        <h2
          class="StyleComponents__Breadcrumb-sc-11erdxt-4 cNjPuO"
        >
          <a
            href="/premiered/movie-detail/undefined"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-angle-left fa-w-8 mr-10"
              data-icon="angle-left"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 256 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
                fill="currentColor"
              />
            </svg>
          </a>
        </h2>
        <div
          class="style__Quantity-wqle2v-0 eTLLsO"
        >
          <p
            class="breadcrumb"
          >
            number of tickets
          </p>
          <div>
            <button
              data-testid="decrement"
            >
              -
            </button>
            <span
              data-testid="ticket-quantity"
            >
              0
            </span>
            <button
              data-testid="increment"
            >
              +
            </button>
          </div>
        </div>
        <button
          class="StyleComponents__Button-sc-11erdxt-3 fWbzoT"
          disabled=""
        >
          <a
            href="/booking/undefined/seat-selection"
          >
            select location
          </a>
        </button>
        <div
          class="StyleComponents__Alert-sc-11erdxt-5 bPYMDT"
        >
          you must select at least one ticket
        </div>
      </div>
    </div>
  </div>
</div>
`;
